path_dynamic_creds.go

```diff
@@ -0,0 +1,242 @@
+		"username": username,
+		"password": password,
+		"dn":       addReq.DN,
+	}
+	internal := map[string]interface{}{
+		"name":     roleName,
+		"username": username,
+		"password": password,
+		"dn":       addReq.DN,
+	}
+	resp := b.Secret(secretCredsType).Response(respData, internal)
+	resp.Secret.TTL = dRole.DefaultTTL
+	resp.Secret.MaxTTL = dRole.MaxTTL
+
+	return resp, nil
+}
+
+func getAddRequest(req *ldif.Entry) *ldap.AddRequest {
+	if req.Add != nil {
+		return req.Add
+	}
+
+	if req.Entry == nil {
+		return nil
+	}
+
+	// Attempt to convert the Entry to an AddRequest
+	attributes := make([]ldap.Attribute, 0, len(req.Entry.Attributes))
+	for _, entryAttribute := range req.Entry.Attributes {
+		attribute := ldap.Attribute{
+			Type: entryAttribute.Name,
+			Vals: entryAttribute.Values,
+		}
+		attributes = append(attributes, attribute)
+	}
+	addReq := &ldap.AddRequest{
+		DN:         req.Entry.DN,
+		Attributes: attributes,
+		Controls:   nil,
+	}
+	return addReq
+}
+
+func (b *backend) secretCredsRenew() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		// Retrieve the role to ensure it still exists. If it doesn't, this will reject the renewal request.
+		roleNameRaw, ok := req.Secret.InternalData["name"]
+		if !ok {
+			return nil, fmt.Errorf("missing role name")
+		}
+
+		roleName := roleNameRaw.(string)
+		dRole, err := retrieveDynamicRole(ctx, req.Storage, roleName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve dynamic role: %w", err)
+		}
+		if dRole == nil {
+			return nil, fmt.Errorf("unable to renew: role does not exist")
+		}
+
+		resp := &logical.Response{
+			Secret: req.Secret,
+		}
+		return resp, nil
+	}
+}
+
+func (b *backend) secretCredsRevoke() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		config, err := readConfig(ctx, req.Storage)
+		if err != nil {
+			return nil, err
+		}
+		if config == nil {
+			return nil, fmt.Errorf("missing OpenLDAP configuration")
+		}
+
+		dn, err := getString(req.Secret.InternalData, "dn")
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve dn from revocation data: %w", err)
+		}
+		if dn == "" {
+			return nil, fmt.Errorf("no DN found in revocation data")
+		}
+
+		delReq := &ldap.DelRequest{
+			DN: dn,
+		}
+
+		err = b.client.Del(config.LDAP, delReq)
+		if err != nil {
+			return nil, fmt.Errorf("failed to revoke OpenLDAP credentials: %w", err)
+		}
+		return nil, nil
+	}
+}
+
+type usernameTemplateData struct {
+	DisplayName string
+	RoleName    string
+}
+
+func generateUsername(req *logical.Request, role *dynamicRole) (string, error) {
+	if role.UsernameTemplate == "" {
+		randStr, err := base62.Random(20)
+		if err != nil {
+			return "", err
+		}
+		username := fmt.Sprintf("v_%s_%s_%s_%d", req.DisplayName, role.Name, randStr, time.Now().Unix())
```

Comment:
Is this default format documented elsewhere?

Reply:
I was planning on including example outputs in the documentation, however I don't plan on explicitly including the format as a guaranteed part of the API. This gives us the potential flexibility to change it and matches the behavior of the database engine.